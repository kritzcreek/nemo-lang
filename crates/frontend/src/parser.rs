use crate::syntax::{AstNode, Root, SyntaxNode};
use rowan::GreenNode;
use std::str;

mod error;
mod grammar;
mod lexer;
mod parsing;

pub use error::ParseError;
pub use lexer::SyntaxKind;

pub fn parse_prog(input: &str) -> Parse {
    let mut p = parsing::Parser::new(input);
    let c = p.checkpoint();
    grammar::prog(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.error("failed to make progress");
        p.start_node(SyntaxKind::Error);
        while !p.at(SyntaxKind::EOF) {
            p.bump_any()
        }
        p.finish_node()
    }
    p.finish_at(c, SyntaxKind::Root);
    let green_node = p.builder.finish();
    Parse {
        green_node,
        errors: p.errors,
    }
}

pub struct Parse {
    green_node: GreenNode,
    errors: Vec<ParseError>,
}

impl Parse {
    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }

    pub fn take(self) -> (Root, Vec<ParseError>) {
        (
            Root::cast(SyntaxNode::new_root(self.green_node)).unwrap(),
            self.errors,
        )
    }
}
