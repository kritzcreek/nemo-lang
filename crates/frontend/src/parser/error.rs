use ariadne::{Color, Config, Label, Report, ReportKind, Source};
use line_index::{LineCol, LineIndex};
use std::{fmt, str};
use text_size::TextRange;

#[derive(Debug)]
pub struct ParseError {
    pub it: String,
    pub at: TextRange,
}

impl ParseError {
    pub fn display<'err, 'src>(
        &'err self,
        source: &'src str,
        colors: bool,
    ) -> ParseErrorDisplay<'src, 'err> {
        ParseErrorDisplay {
            source,
            parse_error: self,
            colors,
        }
    }

    pub fn line_col(&self, line_index: &LineIndex) -> (LineCol, LineCol) {
        let start = line_index.line_col(self.at.start());
        let end = line_index.line_col(self.at.end());
        (start, end)
    }
}

pub struct ParseErrorDisplay<'src, 'err> {
    source: &'src str,
    parse_error: &'err ParseError,
    colors: bool,
}

impl fmt::Display for ParseErrorDisplay<'_, '_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        render_parse_error(self.source, self.parse_error, self.colors, f);
        Ok(())
    }
}

fn render_parse_error(source: &str, error: &ParseError, colors: bool, output: &mut fmt::Formatter) {
    let file_name = "source";

    let out = Color::Fixed(81);
    let cache = (file_name, Source::from(source));

    // TODO avoid this allocation
    let mut out_buf = Vec::new();
    Report::build(ReportKind::Error, file_name, 12)
        .with_code("Parsing error")
        .with_message(&error.it)
        .with_label(
            Label::new((file_name, error.at.start().into()..error.at.end().into()))
                .with_message(error.it.to_string())
                .with_color(out),
        )
        .with_config(Config::default().with_color(colors))
        .finish()
        .write(cache, &mut out_buf)
        .unwrap();

    writeln!(output, "{}", str::from_utf8(&out_buf).unwrap()).unwrap();
}
