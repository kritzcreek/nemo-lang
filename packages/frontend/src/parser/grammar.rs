use rowan::Checkpoint;

use super::Parser;
use crate::lexer::SyntaxKind;
use crate::T;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
pub enum Progress {
    Made,
    None,
}

impl Progress {
    pub fn made_progress(&self) -> bool {
        *self == Progress::Made
    }
}

pub fn prog(p: &mut Parser) {
    while !p.at(SyntaxKind::EOF) {
        if !toplevel(p).made_progress() {
            let c = p.checkpoint();
            // RECOVERY
            p.error("expected a top-level declaration");
            while !p.at(SyntaxKind::EOF) && !TOP_LEVEL_FIRST.contains(&p.current()) {
                p.bump_any();
            }
            p.finish_at(c, SyntaxKind::Error)
        }
    }
}

const TOP_LEVEL_FIRST: [SyntaxKind; 4] = [T![global], T![fn], T![import], T![struct]];
fn toplevel(p: &mut Parser) -> Progress {
    match p.current() {
        T![global] => {
            let c = p.checkpoint();
            p.bump(T![global]);
            p.expect(SyntaxKind::IDENT);
            typ_annot(p);
            p.expect(SyntaxKind::EQUALS);
            if !expr(p).made_progress() {
                p.error("expected an expression")
            }
            p.finish_at(c, SyntaxKind::TopGlobal);
            Progress::Made
        }
        T![fn] => {
            top_fn(p);
            Progress::Made
        }
        T![import] => {
            top_import(p);
            Progress::Made
        }
        T![struct] => {
            top_struct(p);
            Progress::Made
        }
        T![variant] => {
            top_variant(p);
            Progress::Made
        }
        _ => Progress::None,
    }
}

fn top_import(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![import]);
    p.start_node(SyntaxKind::ImpInternal);
    p.expect(T![ident]);
    p.finish_node();
    p.expect(T![:]);
    if !typ(p).made_progress() {
        p.error("expected a function type")
    }
    p.expect(T![from]);
    p.start_node(SyntaxKind::ImpExternal);
    p.expect(T![ident]);
    p.finish_node();
    p.finish_at(c, SyntaxKind::TopImport)
}

fn top_struct(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![struct]);
    p.expect(T![upper_ident]);
    p.expect(T!['{']);
    while !p.at(SyntaxKind::EOF) && !p.at(T!['}']) {
        let c = p.checkpoint();
        p.expect(T![ident]);
        p.expect(T![:]);
        if !typ(p).made_progress() {
            p.error("expected a type")
        }
        p.finish_at(c, SyntaxKind::StructField);

        if !p.at(T!['}']) && !p.expect(T![,]) {
            break;
        }
    }
    p.expect(T!['}']);
    p.finish_at(c, SyntaxKind::TopStruct)
}

fn top_variant(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![variant]);
    p.expect(T![upper_ident]);
    p.expect(T!['{']);
    while !p.at(SyntaxKind::EOF) && !p.at(T!['}']) {
        if !p.at(T![struct]) {
            p.error("expected a struct");
            break;
        }
        top_struct(p);
        if !p.at(T!['}']) && !p.expect(T![,]) {
            break;
        }
    }
    p.expect(T!['}']);
    p.finish_at(c, SyntaxKind::TopVariant)
}

fn top_fn(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(SyntaxKind::FN_KW);
    if !p.eat(SyntaxKind::IDENT) {
        p.error("expected a function name")
    }
    param_list(p);
    if p.eat(T![->]) && !typ(p).made_progress() {
        p.error("expected a return type")
    }
    if !block_expr(p).made_progress() {
        p.error("expected a function body")
    }
    p.finish_at(c, SyntaxKind::TopFn)
}

fn param_list(p: &mut Parser) {
    if !p.eat(SyntaxKind::L_PAREN) {
        p.error("expected a parameter list");
        return;
    }
    while p.at(SyntaxKind::IDENT) {
        let c = p.checkpoint();
        p.bump(SyntaxKind::IDENT);
        if !typ_annot(p).made_progress() {
            p.error("expected a type annotation")
        }
        p.eat(SyntaxKind::COMMA);
        p.finish_at(c, SyntaxKind::Param)
    }
    if !p.eat(SyntaxKind::R_PAREN) {
        p.error("expected a closing paren");
        // TODO recover
    }
}

fn qualifier(p: &mut Parser) -> Progress {
    if p.at(T![upper_ident]) && p.nth_at(1, T![::]) {
        let c = p.checkpoint();
        p.bump(T![upper_ident]);
        p.bump(T![::]);
        p.finish_at(c, SyntaxKind::Qualifier);
        Progress::Made
    } else {
        Progress::None
    }
}

fn typ_annot(p: &mut Parser) -> Progress {
    if !p.eat(SyntaxKind::COLON) {
        return Progress::None;
    }
    if !typ(p).made_progress() {
        p.error("expected a type")
    };
    Progress::Made
}

fn typ(p: &mut Parser) -> Progress {
    let c = p.checkpoint();
    match p.current() {
        T![i32] => {
            p.bump(T![i32]);
            p.finish_at(c, SyntaxKind::TyInt)
        }
        T![f32] => {
            p.bump(T![f32]);
            p.finish_at(c, SyntaxKind::TyFloat)
        }
        T![bool] => {
            p.bump(T![bool]);
            p.finish_at(c, SyntaxKind::TyBool)
        }
        T![unit] => {
            p.bump(T![unit]);
            p.finish_at(c, SyntaxKind::TyUnit)
        }
        T![upper_ident] => {
            qualifier(p);
            p.expect(T![upper_ident]);
            p.finish_at(c, SyntaxKind::TyCons)
        }
        T!['['] => {
            p.bump(T!['[']);
            if !typ(p).made_progress() {
                p.error("expected a type")
            }
            p.expect(T![']']);
            p.finish_at(c, SyntaxKind::TyArray)
        }
        T![fn] => {
            p.bump(T![fn]);
            let c_arg = p.checkpoint();
            p.expect(T!['(']);
            while !p.at(SyntaxKind::EOF) && !p.at(T![')']) {
                typ(p);
                if !p.at(T![')']) && !p.expect(T![,]) {
                    break;
                }
            }
            p.expect(T![')']);
            p.finish_at(c_arg, SyntaxKind::TyArgList);
            p.expect(T![->]);
            if !typ(p).made_progress() {
                p.error("expected a return type")
            }
            p.finish_at(c, SyntaxKind::TyFn)
        }
        _ => return Progress::None,
    }

    Progress::Made
}

// const EXPR_FIRST: [SyntaxKind; 11] = [
//     T![true],
//     T![false],
//     T![int_lit],
//     T![float_lit],
//     T![ident],
//     T![upper_ident],
//     T!['('],
//     T![if],
//     T!['{'],
//     T!['['],
//     T![at_ident],
// ];

fn lit(p: &mut Parser) -> Progress {
    let c = p.checkpoint();
    match p.current() {
        T![true] | T![false] => {
            p.bump_any();
            p.finish_at(c, SyntaxKind::LitBool)
        }
        T![int_lit] | T![binary_lit] | T![hex_lit] => {
            p.bump_any();
            p.finish_at(c, SyntaxKind::LitInt)
        }
        T![float_lit] => {
            p.bump_any();
            p.finish_at(c, SyntaxKind::LitFloat)
        }
        _ => return Progress::None,
    }

    Progress::Made
}

fn expr_bp(p: &mut Parser, min_bp: u32) -> Progress {
    let c = p.checkpoint();
    if !atom(p).made_progress() {
        return Progress::None;
    }
    postfix_expr(p, c);

    loop {
        let Some((op_bp, op)) = current_bin_op(p) else {
            break;
        };

        if op_bp < min_bp {
            break;
        }

        p.start_node(SyntaxKind::BinOp);
        p.bump(op);
        p.finish_node();
        if !expr_bp(p, op_bp).made_progress() {
            p.error("expected an expression");
        }
        p.finish_at(c, SyntaxKind::EBinary)
    }

    Progress::Made
}

fn current_bin_op(p: &mut Parser) -> Option<(u32, SyntaxKind)> {
    let c = p.current();
    match c {
        T![&&] => Some((1, c)),
        T![||] => Some((2, c)),
        T![==] | T![!=] | T![<] | T![>] | T![<=] | T![>=] => Some((3, c)),
        T![+] | T![-] => Some((4, c)),
        T![*] | T![/] => Some((5, c)),
        _ => None,
    }
}

fn expr(p: &mut Parser) -> Progress {
    if p.at(T![if]) {
        if_expr(p);
        return Progress::Made;
    }
    if p.at(T!['{']) {
        return block_expr(p);
    }
    if p.at(T![match]) {
        match_expr(p);
        return Progress::Made;
    }
    expr_bp(p, 0)
}

fn if_expr(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![if]);
    if !expr(p).made_progress() {
        p.error("expected a condition")
    }
    if !block_expr(p).made_progress() {
        p.error("expected a block as the then branch")
    }
    p.expect(T![else]);
    if !block_expr(p).made_progress() {
        p.error("expected a block as the else branch")
    }
    p.finish_at(c, SyntaxKind::EIf)
}

fn block_expr(p: &mut Parser) -> Progress {
    let c = p.checkpoint();
    if !p.eat(T!['{']) {
        return Progress::None;
    }
    while !p.at(SyntaxKind::EOF) && !p.at(T!['}']) {
        if !decl(p).made_progress() {
            break;
        }

        if !p.at(T!['}']) && !p.expect(T![;]) {
            break;
        }
    }
    p.expect(T!['}']);
    p.finish_at(c, SyntaxKind::EBlock);
    Progress::Made
}

fn match_expr(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![match]);
    expr_bp(p, 0);
    p.expect(T!['{']);
    while !p.at(SyntaxKind::EOF) && !p.at(T!['}']) {
        if !match_branch(p).made_progress() {
            break;
        }
        if !p.at(T!['}']) && !p.expect(T![,]) {
            break;
        }
    }
    p.expect(T!['}']);
    p.finish_at(c, SyntaxKind::EMatch)
}

fn match_branch(p: &mut Parser) -> Progress {
    let c = p.checkpoint();
    if !pattern(p).made_progress() {
        return Progress::None;
    }
    p.expect(T![=>]);
    if !block_expr(p).made_progress() {
        p.error("expected a body for this branch")
    }
    p.finish_at(c, SyntaxKind::EMatchBranch);
    Progress::Made
}

fn pattern(p: &mut Parser) -> Progress {
    let c = p.checkpoint();
    match p.current() {
        T![ident] => {
            p.bump(T![ident]);
            p.finish_at(c, SyntaxKind::PatVar);
            Progress::Made
        }
        T![upper_ident] => {
            qualifier(p);
            p.expect(T![upper_ident]);
            p.expect(T![ident]);
            p.finish_at(c, SyntaxKind::PatVariant);
            Progress::Made
        }
        _ => Progress::None,
    }
}

fn call_arg_list(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T!['(']);
    while !p.at(SyntaxKind::EOF) && !p.at(T![')']) {
        if !expr(p).made_progress() {
            break;
        }

        if !p.at(T![')']) && !p.expect(T![,]) {
            break;
        }
    }
    p.expect(T![')']);
    p.finish_at(c, SyntaxKind::EArgList);
}

fn postfix_expr(p: &mut Parser, c: Checkpoint) {
    loop {
        match p.current() {
            T!['('] => {
                call_arg_list(p);
                p.finish_at(c, SyntaxKind::ECall)
            }
            T!['['] => {
                p.bump(T!['[']);
                if !expr(p).made_progress() {
                    p.error("expected an index")
                }
                p.expect(T![']']);
                p.finish_at(c, SyntaxKind::EArrayIdx)
            }
            T![.] => {
                p.bump(T![.]);
                if !p.eat(T![ident]) {
                    p.error("expected a field index")
                }
                p.finish_at(c, SyntaxKind::EStructIdx)
            }
            _ => break,
        }
    }
}

fn atom(p: &mut Parser) -> Progress {
    let c = p.checkpoint();

    if lit(p).made_progress() {
        p.finish_at(c, SyntaxKind::ELit);
        return Progress::Made;
    }

    match p.current() {
        T!['('] => {
            p.bump(T!['(']);
            expr(p);
            p.expect(T![')']);
            p.finish_at(c, SyntaxKind::EParen);
        }
        T!['['] => {
            p.bump(T!['[']);
            while !p.at(SyntaxKind::EOF) && !p.at(T![']']) {
                if !expr(p).made_progress() {
                    break;
                }

                if !p.at(T![']']) && !p.expect(T![,]) {
                    break;
                }
            }

            p.expect(T![']']);
            p.finish_at(c, SyntaxKind::EArray)
        }
        T![upper_ident] => {
            qualifier(p);
            p.expect(T![upper_ident]);
            if p.expect(T!['{']) {
                while !p.at(SyntaxKind::EOF) && !p.at(T!['}']) {
                    let c = p.checkpoint();
                    p.expect(T![ident]);
                    p.expect(T![=]);
                    if !expr(p).made_progress() {
                        p.error("expected a field expression")
                    }
                    p.finish_at(c, SyntaxKind::EStructField);

                    if !p.at(T!['}']) && !p.expect(T![,]) {
                        break;
                    }
                }
                p.expect(T!['}']);
                p.finish_at(c, SyntaxKind::EStruct)
            }
        }
        T![ident] => {
            p.bump(T![ident]);
            p.finish_at(c, SyntaxKind::EVar)
        }
        T![at_ident] => {
            p.bump(T![at_ident]);
            if p.at(T!['(']) {
                call_arg_list(p);
                p.finish_at(c, SyntaxKind::EIntrinsic)
            } else {
                p.error("expected an open parenthesis");
                p.finish_at(c, SyntaxKind::Error)
            }
        }
        _ => return Progress::None,
    }

    Progress::Made
}

fn decl(p: &mut Parser) -> Progress {
    match p.current() {
        T![let] => let_decl(p),
        T![set] => set_decl(p),
        T![while] => while_decl(p),
        _ => {
            let c = p.checkpoint();
            if expr(p).made_progress() {
                p.finish_at(c, SyntaxKind::DExpr)
            } else {
                return Progress::None;
            }
        }
    }

    Progress::Made
}

fn let_decl(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![let]);
    p.expect(T![ident]);
    typ_annot(p);
    p.expect(T![=]);
    if !expr(p).made_progress() {
        p.error("expected an expression")
    }
    p.finish_at(c, SyntaxKind::DLet)
}

fn set_decl(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![set]);
    set_target(p);
    p.expect(T![=]);
    if !expr(p).made_progress() {
        p.error("expected an expression")
    }
    p.finish_at(c, SyntaxKind::DSet)
}

fn set_target(p: &mut Parser) {
    let c = p.checkpoint();
    if p.expect(T![ident]) {
        p.finish_at(c, SyntaxKind::EVar)
    }
    while !p.at(SyntaxKind::EOF) && !p.at(T![=]) {
        match p.current() {
            T![.] | T!['['] => postfix_expr(p, c),
            _ => {
                p.error("expected a set target");
                break;
            }
        }
    }
    p.finish_at(c, SyntaxKind::SetTarget)
}

fn while_decl(p: &mut Parser) {
    let c = p.checkpoint();
    p.bump(T![while]);
    if !expr(p).made_progress() {
        p.error("expected an expression")
    }
    if !block_expr(p).made_progress() {
        p.error("expected a while body")
    }
    p.finish_at(c, SyntaxKind::DWhile)
}
