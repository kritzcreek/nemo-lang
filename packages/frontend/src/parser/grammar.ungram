Root = TopLevel*

// TopLevel
TopLevel
  = TopImport
  | TopGlobal
  | TopStruct
  | TopVariant
  | TopFn

TopImport = 'import' ImpInternal ':' Type 'from' ImpExternal
ImpInternal = 'ident'
ImpExternal = 'ident'

TopGlobal = 'global' 'ident' ':' Type '=' Expr

TopStruct = 'struct' 'upper_ident' '{' StructField* '}'
StructField = 'ident' ':' Type

TopVariant = 'variant' 'upper_ident' '{' TopStruct* '}'

TopFn = 'fn' 'ident' '(' Param* ')' (':' Type)? '=' body:EBlock
Param = 'ident' ':' Type

// Types
Type
  = TyInt
  | TyFloat
  | TyBool
  | TyUnit
  | TyArray
  | TyCons
  | TyFn

TyInt = 'i32'
TyFloat = 'f32'
TyBool = 'bool'
TyUnit = 'unit'
TyArray = '[' elem:Type ']'
TyCons = Qualifier? 'upper_ident'

Qualifier = 'upper_ident' '::'

TyFn = 'fn' TyArgList '->' result:Type
TyArgList = '(' Type* ')'

// Literals
Literal = LitBool | LitFloat | LitInt

LitInt = 'int_lit'
LitBool = 'true' | 'false'
LitFloat = 'float_lit'

// Expressions
Expr
  = ELit
  | EVar
  | EArray
  | EStruct
  | ECall
  | EParen
  | EBinary
  | EArrayIdx
  | EStructIdx
  | EIf
  | EMatch
  | EBlock
  | EIntrinsic

ELit = Literal
EVar = 'ident'
EArray = '[' Expr* ']'
EParen = '(' Expr ')'
EArrayIdx = Expr '[' index:Expr ']'

ECall = Expr EArgList
EArgList = '(' Expr* ')'

EBinary = lhs:Expr op:('+' | '-') rhs:Expr

EStruct = Qualifier? 'upper_ident' '{' EStructField* '}'
EStructField = 'ident' '=' Expr

EStructIdx = Expr '.' 'ident'
EIf = 'if' condition:Expr then_branch:EBlock 'else' else_branch:EBlock

EMatch = 'match' scrutinee:Expr '{' EMatchBranch* '}'
EMatchBranch = Pattern '=>' body:EBlock

EBlock = '{' Declaration* '}'

EIntrinsic = 'at_ident' EArgList

// Declarations
Declaration
  = DLet
  | DSet
  | DWhile
  | DExpr

DLet = 'let' 'ident' (':' Type)? '=' Expr
DSet = 'set' SetTarget '=' Expr
DWhile = 'while' Expr EBlock
DExpr = Expr

SetTarget = SetTargetExpr
SetTargetExpr = EVar | EArrayIdx | EStructIdx

Pattern
  = PatVariant
  | PatVar

PatVariant = Qualifier 'upper_ident' 'ident'
PatVar = 'ident'
