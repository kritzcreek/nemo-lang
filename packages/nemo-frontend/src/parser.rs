use tree_sitter::{Parser, Tree};
use tree_sitter_highlight::HighlightConfiguration;
use tree_sitter_highlight::HighlightEvent;
use tree_sitter_highlight::Highlighter;

pub const HIGHLIGHT_NAMES: [&'static str; 7] = [
    "keyword", "type", "function", "operator", "property", "number", "comment",
];

pub fn parse_program(program: &str) -> Tree {
    let mut parser = Parser::new();
    parser.set_language(tree_sitter_nemo::language()).unwrap();
    parser.parse(program, None).unwrap()
}

pub fn highlight(program: &str) -> Vec<HighlightEvent> {
    let nemo_language = tree_sitter_nemo::language();

    let mut nemo_config =
        HighlightConfiguration::new(nemo_language, tree_sitter_nemo::HIGHLIGHTS_QUERY, "", "")
            .unwrap();

    nemo_config.configure(&HIGHLIGHT_NAMES);

    let mut highlighter = Highlighter::new();

    let highlights = highlighter
        .highlight(&nemo_config, program.as_bytes(), None, |_| None)
        .unwrap();

    highlights.map(|e| e.unwrap()).collect()
}
