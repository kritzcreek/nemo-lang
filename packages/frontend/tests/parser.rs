use frontend::parser::parse_prog;
use insta::{assert_snapshot, glob};
use std::fmt::Write;
use std::str;

/// Copied from the Rust compiler: https://github.com/rust-lang/rust/pull/62948/
/// Replaces `\r\n` with `\n` in-place in `src`.
///
/// Returns error if there's a lone `\r` in the string
fn normalize_newlines(src: &mut String) {
    if !src.as_bytes().contains(&b'\r') {
        return;
    }

    // We replace `\r\n` with `\n` in-place, which doesn't break utf-8 encoding.
    // While we *can* call `as_mut_vec` and do surgery on the live string
    // directly, let's rather steal the contents of `src`. This makes the code
    // safe even if a panic occurs.

    let mut buf = std::mem::replace(src, String::new()).into_bytes();
    let mut gap_len = 0;
    let mut tail = buf.as_mut_slice();
    loop {
        let idx = match find_crlf(&tail[gap_len..]) {
            None => tail.len(),
            Some(idx) => idx + gap_len,
        };
        tail.copy_within(gap_len..idx, 0);
        tail = &mut tail[idx - gap_len..];
        if tail.len() == gap_len {
            break;
        }
        gap_len += 1;
    }

    // Account for removed `\r`.
    // After `set_len`, `buf` is guaranteed to contain utf-8 again.
    let new_len = buf.len() - gap_len;
    unsafe {
        buf.set_len(new_len);
        *src = String::from_utf8_unchecked(buf);
    }

    fn find_crlf(src: &[u8]) -> Option<usize> {
        let mut search_idx = 0;
        while let Some(idx) = find_cr(&src[search_idx..]) {
            if src[search_idx..].get(idx + 1) != Some(&b'\n') {
                search_idx += idx + 1;
                continue;
            }
            return Some(search_idx + idx);
        }
        None
    }

    fn find_cr(src: &[u8]) -> Option<usize> {
        src.iter().position(|&b| b == b'\r')
    }
}

#[test]
fn parse_toplevel() {
    use std::fs;

    glob!("passing/*.nemo", |path| {
        let input = {
            let mut input = fs::read_to_string(path).unwrap();
            normalize_newlines(&mut input);
            input
        };
        for inp in input.split("// ---\n") {
            let parse = parse_prog(inp);
            if parse.has_errors() {
                eprintln!(
                    "{}: Failed on {:?} with {:?}",
                    path.display(),
                    inp,
                    parse.errors
                );
            }
            let output = format!("{}\n// ---\n{}", inp, parse.debug_tree());
            assert_snapshot!(output);
        }
    });
}

#[test]
fn parse_example() {
    use std::fs;
    let input = {
        let mut input = fs::read_to_string("../../example.nemo").unwrap();
        normalize_newlines(&mut input);
        input
    };
    let mut error_output = String::new();
    let parse = parse_prog(&input);
    if parse.has_errors() {
        writeln!(&mut error_output, "=== ERRORS ===").unwrap();
        for error in &parse.errors {
            writeln!(&mut error_output, "{}", error.display(&input, false)).unwrap();
        }
    }
    let output = format!("{}\n{}", parse.debug_tree(), error_output);
    assert_snapshot!(output);
}
