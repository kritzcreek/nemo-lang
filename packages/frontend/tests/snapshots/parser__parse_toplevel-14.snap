---
source: packages/frontend/tests/parser.rs
expression: output
input_file: packages/frontend/tests/passing/toplevel.nemo
---
import internal : fn (i32, f32) -> i32 from external

struct Vec {
  x : f32,
  y : f32,
}

global x = 10
global y : f32 = 10.0

fn main() = {

}

fn add(x : i32, y : i32) : i32 = {
    x + y
}


// ---
Root@0..194
  TopImport@0..53
    LINEFEED@0..1 "\n"
    IMPORT_KW@1..7 "import"
    SPACE@7..8 " "
    ImpInternal@8..17
      IDENT@8..16 "internal"
      SPACE@16..17 " "
    COLON@17..18 ":"
    SPACE@18..19 " "
    TyFn@19..40
      FN_KW@19..21 "fn"
      SPACE@21..22 " "
      TyArgList@22..33
        L_PAREN@22..23 "("
        TyInt@23..26
          I32_BUILTIN@23..26 "i32"
        COMMA@26..27 ","
        SPACE@27..28 " "
        TyFloat@28..31
          F32_BUILTIN@28..31 "f32"
        R_PAREN@31..32 ")"
        SPACE@32..33 " "
      ARROW@33..35 "->"
      SPACE@35..36 " "
      TyInt@36..40
        I32_BUILTIN@36..39 "i32"
        SPACE@39..40 " "
    FROM_KW@40..44 "from"
    SPACE@44..45 " "
    ImpExternal@45..53
      IDENT@45..53 "external"
  TopStruct@53..91
    LINEFEED@53..54 "\n"
    LINEFEED@54..55 "\n"
    STRUCT_KW@55..61 "struct"
    SPACE@61..62 " "
    UPPER_IDENT@62..65 "Vec"
    SPACE@65..66 " "
    L_BRACE@66..67 "{"
    StructField@67..77
      LINEFEED@67..68 "\n"
      SPACE@68..70 "  "
      IDENT@70..71 "x"
      SPACE@71..72 " "
      COLON@72..73 ":"
      SPACE@73..74 " "
      TyFloat@74..77
        F32_BUILTIN@74..77 "f32"
    COMMA@77..78 ","
    StructField@78..88
      LINEFEED@78..79 "\n"
      SPACE@79..81 "  "
      IDENT@81..82 "y"
      SPACE@82..83 " "
      COLON@83..84 ":"
      SPACE@84..85 " "
      TyFloat@85..88
        F32_BUILTIN@85..88 "f32"
    COMMA@88..89 ","
    LINEFEED@89..90 "\n"
    R_BRACE@90..91 "}"
  TopGlobal@91..106
    LINEFEED@91..92 "\n"
    LINEFEED@92..93 "\n"
    GLOBAL_KW@93..99 "global"
    SPACE@99..100 " "
    IDENT@100..101 "x"
    SPACE@101..102 " "
    EQUALS@102..103 "="
    SPACE@103..104 " "
    ELit@104..106
      LitInt@104..106
        INT_LIT@104..106 "10"
  TopGlobal@106..128
    LINEFEED@106..107 "\n"
    GLOBAL_KW@107..113 "global"
    SPACE@113..114 " "
    IDENT@114..115 "y"
    SPACE@115..116 " "
    COLON@116..117 ":"
    SPACE@117..118 " "
    TyFloat@118..122
      F32_BUILTIN@118..121 "f32"
      SPACE@121..122 " "
    EQUALS@122..123 "="
    SPACE@123..124 " "
    ELit@124..128
      LitFloat@124..128
        FLOAT_LIT@124..128 "10.0"
  TopFn@128..146
    LINEFEED@128..129 "\n"
    LINEFEED@129..130 "\n"
    FN_KW@130..132 "fn"
    SPACE@132..133 " "
    IDENT@133..137 "main"
    L_PAREN@137..138 "("
    R_PAREN@138..139 ")"
    SPACE@139..140 " "
    EQUALS@140..141 "="
    SPACE@141..142 " "
    EBlock@142..146
      L_BRACE@142..143 "{"
      LINEFEED@143..144 "\n"
      LINEFEED@144..145 "\n"
      R_BRACE@145..146 "}"
  TopFn@146..194
    LINEFEED@146..147 "\n"
    LINEFEED@147..148 "\n"
    FN_KW@148..150 "fn"
    SPACE@150..151 " "
    IDENT@151..154 "add"
    L_PAREN@154..155 "("
    Param@155..164
      IDENT@155..156 "x"
      SPACE@156..157 " "
      COLON@157..158 ":"
      SPACE@158..159 " "
      TyInt@159..162
        I32_BUILTIN@159..162 "i32"
      COMMA@162..163 ","
      SPACE@163..164 " "
    Param@164..171
      IDENT@164..165 "y"
      SPACE@165..166 " "
      COLON@166..167 ":"
      SPACE@167..168 " "
      TyInt@168..171
        I32_BUILTIN@168..171 "i32"
    R_PAREN@171..172 ")"
    SPACE@172..173 " "
    COLON@173..174 ":"
    SPACE@174..175 " "
    TyInt@175..179
      I32_BUILTIN@175..178 "i32"
      SPACE@178..179 " "
    EQUALS@179..180 "="
    SPACE@180..181 " "
    EBlock@181..194
      L_BRACE@181..182 "{"
      DExpr@182..192
        EBinary@182..192
          EVar@182..189
            LINEFEED@182..183 "\n"
            SPACE@183..187 "    "
            IDENT@187..188 "x"
            SPACE@188..189 " "
          BinOp@189..191
            PLUS@189..190 "+"
            SPACE@190..191 " "
          EVar@191..192
            IDENT@191..192 "y"
      LINEFEED@192..193 "\n"
      R_BRACE@193..194 "}"
